types/decimal.go

```diff
@@ -0,0 +1,403 @@
+	quo, rem = quo.QuoRem(d, precisionReuse, rem)
+
+	if rem.Sign() == 0 { // remainder is zero
+		return quo
+	}
+
+	//lenWhole := len(d.String())
+	//if quo.Sign() == 0 { // only the decimal places (ex. 0.1234)
+	//lenWhole++
+	//}
+	//lenQuo := len(quo.String())
+	//lenRem := len(rem.String())
+	//leadingZeros := lenWhole - (lenQuo + lenRem) // leading zeros removed from the remainder
+
+	//zerosToAdd := int64(lenRem - 1 + leadingZeros)
+	//multiplier := new(big.Int).Exp(big.NewInt(10), big.NewInt(zerosToAdd), nil)
+	//fiveLine := new(big.Int).Mul(big.NewInt(5), multiplier)
+
+	switch rem.Cmp(fivePrecision) {
+	case -1:
+		chopped = quo
+		return
+	case 1:
+		chopped = quo.Add(quo, big.NewInt(1))
+		return
+	default: // bankers rounding must take place
+		// always round to an even number
+		if quo.Bit(0) == 0 {
+			chopped = quo
+			return
+		}
+		chopped = quo.Add(quo, big.NewInt(1))
+		return
+	}
+}
+
+// RoundInt64 rounds the decimal using bankers rounding
+func (d Dec) RoundInt64() int64 {
+	chopped := ChopPrecisionAndRound(d.Int)
+	if !chopped.IsInt64() {
+		panic("Int64() out of bound")
+	}
+	return chopped.Int64()
+}
+
+// RoundInt round the decimal using bankers rounding
+func (d Dec) RoundInt() Int {
+	return NewIntFromBigInt(ChopPrecisionAndRound(d.Int))
+}
+
+//___________________________________________________________________________________
+
+// wraps d.MarshalText()
+func (d Dec) MarshalAmino() (string, error) {
+	if d.Int == nil {
+		d.Int = new(big.Int)
+	}
+	bz, err := d.Int.MarshalText()
+	return string(bz), err
+}
+
+// requires a valid JSON string - strings quotes and calls UnmarshalText
+func (d *Dec) UnmarshalAmino(text string) (err error) {
+	tempInt := new(big.Int)
+	err = tempInt.UnmarshalText([]byte(text))
+	if err != nil {
+		return err
+	}
+	d.Int = tempInt
+	return nil
+}
+
+// MarshalJSON defines custom encoding scheme
+func (d Dec) MarshalJSON() ([]byte, error) {
+	if d.Int == nil {
+		d.Int = new(big.Int)
+	}
+
+	text, err := d.Int.MarshalText()
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(string(text))
+}
+
+// UnmarshalJSON defines custom decoding scheme
+func (d *Dec) UnmarshalJSON(bz []byte) error {
+	if d.Int == nil {
+		d.Int = new(big.Int)
+	}
+
+	var text string
+	err := json.Unmarshal(bz, &text)
+	if err != nil {
+		return err
+	}
+	return d.Int.UnmarshalText([]byte(text))
+}
+
+//___________________________________________________________________________________
+// helpers
+
+// test if two decimal arrays are equal
+func DecsEqual(d1s, d2s []Dec) bool {
+	if len(d1s) != len(d2s) {
+		return false
+	}
+
+	for i, d1 := range d1s {
+		if !d1.Equal(d2s[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+// minimum decimal between two
+func MinDec(d1, d2 Dec) Dec {
+	if d1.LT(d2) {
+		return d1
+	}
+	return d2
+}
+
+// intended to be used with require/assert:  require.True(DecEq(...))
+func DecEq(t *testing.T, exp, got Dec) (*testing.T, bool, string, Dec, Dec) {
```

Comment:
maybe move it to `_test.go` file?

Reply:
cannot, see: https://github.com/cosmos/cosmos-sdk/pull/1819#discussion_r209438898